server/auth/types/openid/routes.ts

```diff
@@ -0,0 +1,213 @@
+
+export class OpenIdAuthRoutes {
+  constructor(
+    private readonly router: IRouter,
+    private readonly config: SecurityPluginConfigType,
+    private readonly sessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>,
+    private readonly openIdAuthConfig: OpenIdAuthConfig,
+    private readonly securityClient: SecurityClient,
+    private readonly core: CoreSetup
+  ) {}
+
+  public setupRoutes() {
+    this.router.get(
+      {
+        path: `/auth/openid/login`,
+        validate: {
+          query: schema.object({
+            code: schema.maybe(schema.string()),
+            nextUrl: schema.maybe(schema.string()),
+            state: schema.maybe(schema.string()),
+            refresh: schema.maybe(schema.string()),
+          }),
+        },
+        options: {
+          authRequired: false,
+        },
+      },
+      async (context, request, response) => {
+        // implementation refers to https://github.com/hapijs/bell/blob/master/lib/oauth.js
+
+        // Sign-in initialization
+        if (!request.query.code) {
+          const nonce = randomString(22);
+          const query: any = {
+            client_id: this.config.openid?.client_id,
+            response_type: 'code',
+            redirect_uri: `${this.getBaseRedirectUrl()}`, // TODO: generate actual domain name
+            state: nonce,
+          };
+
+          const scope = this.config.openid?.scope || '';
+          if (scope) {
+            query.scope = scope;
+          }
+
+          const queryString = stringify(query);
+          const location = `${this.openIdAuthConfig.authorizationEndpoint}?${queryString}`;
+          const cookie: SecuritySessionCookie = {
+            oidcState: nonce,
+          };
+          this.sessionStorageFactory.asScoped(request).set(cookie);
+          return response.redirected({
+            headers: {
+              location: location,
+            },
+          });
+        }
+
+        // Authentication callback
+        // TODO: figure out a decent way to validate the state from cookie
+        const clientId = this.config.openid?.client_id;
+        const clientSecret = this.config.openid?.client_secret;
+        const query: any = {
+          grant_type: 'authorization_code',
+          code: request.query.code,
+          redirect_uri: `${this.getBaseRedirectUrl()}`,
+          client_id: clientId,
+          client_secret: clientSecret,
+        };
+
+        const requestOptions: any = {
+          payload: stringify(query),
+          headers: {
+            'Content-Type': 'application/x-www-form-urlencoded',
+          },
+        };
+
+        try {
+          const tokenResponse = await wreck.post(
+            this.openIdAuthConfig.tokenEndpoint as string,
+            requestOptions
+          );
+          if (
+            !tokenResponse.res.statusCode ||
+            tokenResponse.res.statusCode < 200 ||
+            tokenResponse.res.statusCode > 299
```

Comment:
Minor: personal preference to create a small function for such multi condition ifs for better readability. I will leave it up to you to decide.